### PR TITLE
EVG-8090 refactor type converter code

### PR DESCRIPTION
**Pull request by john-m-liu**
Sorry the pr doesn't really match the ticket because making this work for vendored packages is exactly the same as non-vendored ones. The main changes here are:
- Change the previous logic that had static code to convert between types and pointers to those types to instead be generated, using the model fields to figure out what types it needs to generated
- Add handling for special scalars - Time, Map, Any

### CODE REVIEW

File: rest/model/codegen.go [hunk 1]
```diff
@@ -63,84 +63,113 @@ type modelConversionInfo struct {
 // ModelMapping maps schema type names to their respective DB model
 type ModelMapping map[string]string
 
-func Codegen(schema string, config ModelMapping) ([]byte, error) {
+func Codegen(schema string, config ModelMapping) ([]byte, []byte, error) {
```
Comment (ablack12) [inline rest/model/codegen.go:66]: This function should have a comment explaining the return types so future programmers don't get confused

File: rest/model/codegen.go [hunk 2]
```diff
@@ -63,84 +63,113 @@ type modelConversionInfo struct {
...
-			fieldInfo := getFieldInfo(field)
-			extractedFields[field.Name] = fieldInfo
-			var fieldData string
-			fieldData, err = output(fieldTemplate, fieldInfo)
-			if err != nil {
-				catcher.Add(err)
-				continue
-			}
-			fields += fieldData
+		dbModel, shouldConvert := config[typeName]
```
Comment (ablack12) [inline rest/model/codegen.go:108]: why did you move this to be after gqlType things? If it's not to catch the error at line 99 then you may as well short circuit

File: rest/model/codegen.go [hunk 3]
```diff
@@ -63,84 +63,113 @@ type modelConversionInfo struct {
...
 	structs := ""
-	conversionCode := ""
+	serviceCode := ""
+	generatedConversions := map[string]string{}
 
 	typeNames := []string{}
 	for key := range parsedAst.Types {
 		typeNames = append(typeNames, key)
 	}
 	sort.Strings(typeNames)
```
Comment (ablack12) [inline rest/model/codegen.go:95]: What's the reason for making the list of keys sorted, just good organization?

File: rest/model/codegen.go [hunk 4]
```diff
@@ -63,84 +63,113 @@ type modelConversionInfo struct {
...
 			catcher.Add(err)
 			continue
 		}
 		structs += structData
+		serviceCode += code
+	}
+	modelFile, err := output(fileTemplate, fileInfo{Package: "model", Structs: structs, Code: serviceCode})
+	catcher.Add(err)
+	formattedModelCode, err := goimports(modelFile)
+	catcher.Add(err)
```
Comment (ablack12) [inline rest/model/codegen.go:124]: Can this error be wrapped? The error might not be very pretty

File: rest/model/codegen.go [hunk 5]
```diff
@@ -63,84 +63,113 @@ type modelConversionInfo struct {
...
 	}
-	file, err := output(fileTemplate, fileInfo{Package: "model", Structs: structs, Code: conversionCode})
+	sort.Strings(converters)
+	for _, converterType := range converters {
+		conversionCode += generatedConversions[converterType] + "\n"
+	}
+	converterFile, err := output(fileTemplate, fileInfo{Package: "model", Code: conversionCode})
 	catcher.Add(err)
-	formatted, err := goimports(file)
+	formattedConverterFile, err := goimports(converterFile)
```
Comment (ablack12) [inline rest/model/codegen.go:137]: wrap this error too

File: rest/model/codegen.go [hunk 6]
```diff
@@ -63,84 +63,113 @@ type modelConversionInfo struct {
...
-			}
-			fields += fieldData
+		dbModel, shouldConvert := config[typeName]
+		if !shouldConvert {
+			continue
 		}
-		var structData string
-		structData, err = output(structTemplate, structInfo{Name: typeName, Fields: fields})
+
+		structData, code, err := generateForStruct(fieldTemplate, structTemplate, typeName, dbModel, *gqlType, generatedConversions)
```
Comment (ablack12) [inline rest/model/codegen.go:113]: Doe this destructively add to generatedConversions? That seems shady

File: rest/model/codegen.go [hunk 7]
```diff
@@ -63,84 +63,113 @@ type modelConversionInfo struct {
...
 	catcher.Add(err)
-	formatted, err := goimports(file)
+	formattedConverterFile, err := goimports(converterFile)
 	catcher.Add(err)
 
-	return formatted, catcher.Resolve()
+	return formattedModelCode, formattedConverterFile, catcher.Resolve()
+}
+
+func generateForStruct(fieldTemplate, structTemplate *template.Template, typeName, dbModel string, gqlType ast.Definition, generatedConversions map[string]string) (string, string, error) {
```
Comment (ablack12) [inline rest/model/codegen.go:143]: also a comment about return value here

File: rest/model/codegen.go [hunk 8]
```diff
@@ -63,84 +63,113 @@ type modelConversionInfo struct {
...
+	fields := ""
+	extractedFields := extractedFields{}
+	for _, field := range gqlType.Fields {
+		fieldInfo := getFieldInfo(field)
+		extractedFields[field.Name] = fieldInfo
+		var fieldData string
```
Comment (ablack12) [inline rest/model/codegen.go:149]: nit: don't need this definition anymore

File: rest/model/conversions.go [hunk 9]
```diff
@@ -1,84 +1,95 @@
...
-	if in == nil {
-		return 0
+func noTransforms(typeName string, outIsPtr bool, generatedConversions map[string]string) (string, error) {
+	if err := generateConversionCode(generatedConversions, typeName); err != nil {
+		return "", err
 	}
-	return *in
+	typeName = cleanName(typeName)
+	if outIsPtr {
+		return fmt.Sprintf("%s%sPtr", typeName, typeName), nil
```
Comment (ablack12) [inline rest/model/conversions.go:67]: Will the in type and out type here always match? In which case do we need to repeat it?

File: rest/model/codegen.go [hunk 10]
```diff
@@ -63,84 +63,113 @@ type modelConversionInfo struct {
...
-			return nil, errors.Errorf("invalid format for DB model: %s", dbModel)
-		}
-		code, err := createConversionMethods(strings.Join(parts[:len(parts)-1], "."), parts[len(parts)-1], extractedFields)
-		if err != nil {
-			return nil, errors.Wrapf(err, "error generating conversion methods for type '%s'", typeName)
-		}
-		conversionCode += code
+	converters := []string{}
+	conversionCode := ""
+	for key := range generatedConversions {
```
Comment (ablack12) [inline rest/model/codegen.go:128]: You've done this block a couple times, maybe a util "GetSortedKeys" would be more readable

File: rest/model/codegen.go (hunk 4 above)
Comment (john-m-liu) [reply rest/model/codegen.go:124]: done

File: rest/model/codegen.go (hunk 10 above)
Comment (john-m-liu) [reply rest/model/codegen.go:128]: I tried that previously, but it'll require multiple functions with different signatures since the inputs are map[string]Something so it won't make things cleaner

File: rest/model/codegen.go (hunk 5 above)
Comment (john-m-liu) [reply rest/model/codegen.go:137]: done

File: rest/model/codegen.go (hunk 7 above)
Comment (john-m-liu) [reply rest/model/codegen.go:143]: done

File: rest/model/codegen.go (hunk 8 above)
Comment (john-m-liu) [reply rest/model/codegen.go:149]: done

File: rest/model/conversions.go (hunk 9 above)
Comment (john-m-liu) [reply rest/model/conversions.go:67]: This function is only called in scenarios where the in and out types match

File: rest/model/codegen.go (hunk 2 above)
Comment (ablack12) [reply rest/model/codegen.go]: haha well if you do find a good reason in future tickets let's drop a commend so we remember what's going on
